File: 22/generateParenthesis.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []
        stack = []

        def backtrack(startN, endN):
            if startN == endN == n:
                res.append(''.join(stack))
                return
            
            if startN < n:
                stack.append('(')
                backtrack(startN + 1, endN)
                stack.pop()

            if endN < startN:
                stack.append(')')
                backtrack(startN, endN + 1)
                stack.pop()

        backtrack(0, 0)

        return res

File: 22/generateParenthesis.js
class Solution {
  /**
   * @param {number} openN
   * @param {number} closeN
   * @param {number} n
   * @param {string[]} res
   * @param {string} stack
   */
  backtrack(openN, closedN, n, res, stack) {
    if (openN === closedN && openN === n) {
      res.push(stack);
      return;
    }

    if (openN < n) {
      this.backtrack(openN + 1, closedN, n, res, stack + "(");
    }
    if (closedN < openN) {
      this.backtrack(openN, closedN + 1, n, res, stack + ")");
    }
  }

  /**
   * @param {number} n
   * @return {string[]}
   */
  generateParenthesis(n) {
    const res = [];
    this.backtrack(0, 0, n, res, "");
    return res;
  }
}
